fboom_me.sh

```diff
@@ -0,0 +1,201 @@
+        RESP=$(captcha_process "$IMG_FILE" fboom_me 6 7) || return
+        { read WORD; read ID; } <<< "$RESP"
+        rm -f "$IMG_FILE"
+
+        CAPTCHA_DATA="-d CaptchaForm%5Bcode%5D=$WORD"
+
+        log_debug "Captcha data: $CAPTCHA_DATA"
+
+        FILE_ID=$(parse_form_input_by_name 'uniqueId' <<< "$PAGE") || return
+
+        PAGE=$(curl -b "$COOKIE_FILE" $CAPTCHA_DATA \
+            -d 'free=1' -d 'freeDownloadRequest=1' \
+            -d "uniqueId=$FILE_ID" \
+            -H 'X-Requested-With: XMLHttpRequest' \
+            "$URL") || return
+
+        if match 'The verification code is incorrect' "$PAGE"; then
+            log_error 'Wrong captcha'
+            captcha_nack "$ID"
+            return $ERR_CAPTCHA
+        fi
+
+        log_debug 'Correct captcha'
+        captcha_ack "$ID"
+
+        # parse wait time
+        WAIT=$(parse ' id="download-wait-timer"' \
+            '.*>[[:blank:]]*\([[:digit:]]\+\)' 1 <<< "$PAGE") || return
+
+        if [ -n "$WAIT" ]; then
+            wait $(( WAIT + 1 )) || return
+        fi
+
+        PAGE=$(curl -b "$COOKIE_FILE" -d "uniqueId=$FILE_ID" \
+            -d 'free=1' "$URL") || return
+
+        PRE_URL=$(parse_attr 'id="temp-link"' 'href' <<< "$PAGE") || return
+
+    # direct download without captcha
+    elif match 'bottomDownloadButton' "$PAGE"; then
+        PRE_URL=$(parse 'window.location.href' "'\([^']\+\)'" <<< "$PAGE") || return
+    else
+        log_error 'Unexpected content. Site updated?'
+        return $ERR_FATAL
+    fi
+
+    log_debug "Pre-URL: '$PRE_URL'"
+    PAGE=$(curl --head -b "$COOKIE_FILE" "$BASE_URL$PRE_URL") || return
+
+    # output final url + file name
+    grep_http_header_location <<< "$PAGE" || return
+    echo "$FILE_NAME"
+}
+
+# Probe a download URL
+# $1: cookie file (unused here)
+# $2: fboom.me url
+# $3: requested capability list
+# stdout: 1 capability per line
+#
+# probable API (see this file's header) does not provide an anonymous check-link feature :(
+# $ curl --data '{"ids"=["816bef5d35245"]}' http://fboom.me/api/v1/GetFilesInfo
+fboom_me_probe() {
+    local -r URL=$2
+    local -r REQ_IN=$3
+    local PAGE REQ_OUT FILE_NAME
+
+    PAGE=$(curl --location "$URL") || return
+
+    # File not found or delete
+    if match 'File not found or deleted\|This file is no longer available' "$PAGE"; then
+        return $ERR_LINK_DEAD
+    fi
+
+    REQ_OUT=c
+
+    if [[ $REQ_IN = *f* ]]; then
+        echo "$PAGE" | parse 'Download file:' '^[[:blank:]]\+\([^<]\+\)[[:blank:]]\+<' 3 | \
+        sed -e 's/[[:space:]]*$//' && \
+            REQ_OUT="${REQ_OUT}f"
+    fi
+
+    if [[ $REQ_IN = *s* ]]; then
+        FILE_SIZE=$(parse_tag 'File size:' 'div' <<< "$PAGE") && \
+            translate_size "${FILE_SIZE#File size: }" && REQ_OUT="${REQ_OUT}s"
+    fi
+
+    if [[ $REQ_IN = *i* ]]; then
+        parse . 'file/\([[:alnum:]]\+\)/\?' <<< "$URL" && REQ_OUT="${REQ_OUT}i"
+    fi
+
+    echo $REQ_OUT
+}
```

Comment:
And one more thing that I didn't notice earlier. Modules should contain the "newline" character at the end of the last line. This should not to be confused with a "new line". Vim editor do it automatically, I don't know how other editors do. Nevertheless you can fix it by executing the following sed command:

```
sed -i -e '$a\' fboom_me.sh
```